README
Introduction
============

Badukiplus 0.1 is a C++ rewrite of Baduki, a sophisticated Go program
written by Lim Jaebum. Baduki's origins trace back to AmiGo, which was
itself a C rewrite of a Pascal Go player written by Stoney Ballard.

I already rewrote Wally and Amigo, and someday I plan on writing my own
Go engine in an OO language. In the meantime, I was curious to see how
different Go engines were written, and how strong or weak these simple Go
engines would be compared to recent versions of Gnu Go.

Check Wallyplus and Amigoplus first! They are now on ibiblio.org's Linux
archive.

Baduki is a larger program than Wally and Amigo; it incorporates both a
sophisticated pattern matching algorithm and the complex heuristics found
in Amigo. Again, the data structures are similar.

I rewrote Baduki directly from the original source, version 0.2.9. Basically
I removed all the X Window and GTK related code and added a GTP version 2
compliant front-end. I plan on writing a nice KDE-based GTP V2 client, Real
Soon Now. I work using KDevelop, a nice IDE for KDE.

Below is the original README file from Baduki.

The original Baduki Version 0.2.9 tarball is included in the Documentation
directory.

Andrew D. Balsa <andrebalsa@mailingaddress.org>
13 October 2002

-----------------------------------------------------------------

  Baduki 0.2

  Baduki is a computer Baduk(Go) program.

  Baduki is registed as BadukI on IGS.
  Baduki Homepage: http://soback.kornet21.net/~artist/baduk/baduki.html


Copyright and Acknowledgement:

  Baduki: Copyright(C) 1998-1999 Lim Jaebum (artist@soback.kornet21.net)
  Baduki is Free software.  you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation - version 2.

  Baduki is based on Xamigo( Copyright(C) 1992 Neil Bowers)
and Amigo(Copyright(C) 1989 Todd R. Johnson).
  Gtkgoboard widget: Copyright (C) 1998 Changwoo Ryu.
  Gmp library: Copyright (C) 1995-1997 William Shubert.
     He is author of Cgoban(http://www.hevanet.com/wms/comp/cgoban/)


Features:
  Play Baduk(Go) with Computer or Human.
  Load/Save/Analyze SGF file.
  Suport GMP(go modem protocol)
  Easy to debug.


Install: (Read INSTALL)

  Do the following from the baduki directory:
$ ./configure
$ make
$ make install
$ baduki

  Baduki requires gtk+ 1.0.x.
  Tested Under x86/Alpha Linux libc5/glibc2.

Usage:

-sgf filename		Analyzing sgf file
# ./baduki -sgf in.sgf > out.sgf

-lsgf filename [movenum]	Load sgf file
# ./baduki -lsgf foo.sgf 12

* Baduki automatically saves after each moves and end of game.

-gmp			Gmp mode
With this Baduki is ran under CGoban.
On CGoban's Go Modem menu, type './baduki -gmp' on Program entry.


Feel free to mail me any comments and suggestions for improvements.

Special Thanks to
Ryu Changwoo <cwryu@adam.kaist.ac.kr>
	gtkgoboard and using automake/autoconf/gettext.

--
Artist (artist@soback.kornet21.net)

-----------------------------------------------------------------

How to add a new Pattern
========================

To add a new pattern edit src/patterns.dat.

Symbols in patterns.dat
=======================
  NOTE: Symbols must placed in begin of line.

;	Comment
#	Begin of pattern
	#Idnum Default_value [x_position y_position]
#.	End of patterns
>	Begin of Code for pattern
|	Pattern TYPE
	|PT_CENTER	Check for center only (Default)
	|PT_WHOLE	Check for entire board
	|PT_SIDE	Check for side only
	|PT_CORNER	Check for corder only


Symbols in pattern
==================
  NOTE: 'My' means current turn's, 'His' means opponent's

.	Empty point
O	My stone
o	My stone or empty
X	His stone
x	His stone or empty
S	Empty && My territory from scoreBoard
s	Empty && His territory from scoreBoard
T	Empty && My forcefield from claim
t	Empty && His forcefield from claim
%	Not empty
@	My next move


Coordinate on Pattern
=====================

 | (y,0) ... (x, y)
 |  ..         ..
 | (0,0) ... (x, 0)
 |   -- This side is border if Pattern type is PT_SIDE
 +- Here is Corner if Pattern type is PT_CORNER


Pre-defined Macros for Code
===========================

#define PXY(a, b)       get_px(x, a, b, pat, dir)][get_py(y, a, b, pat, dir)
	Get real position on board for array
	Usage:	INTBOARD[PXY(a,b)] --> INTBOARD[absX][absY]

#define PCXY(a, b)      get_px(x, a, b, pat, dir), get_py(y, a, b, pat, dir)
	Get real position on board for function
	Usage:	FNCTION(PCXY(a,b)) --> FUNCTION(absX, absY)

STRING(a, b)
	string on (a,b) from stringRec sList[]
	
LIBC(a, b)
	Liberty count for string on (a,b)

LIBB(a, b)      libBoard[PXY(a,  b)]

BORD(a, b)      bord[PXY(a, b)]
	0: empty, 1: my stone, -1: his stone

SCORE(a, b)     scoreBoard[PXY(a, b)]
	0: unknown
	2: My Territory
	1: Border of My Territory
	-1, -2: His....
	
CLAIM(a, b)     claim[PXY(a, b)]
	Positive: My force field

POW(a, b)       powerBoard[PXY(a,b)]
	How many claim[] changes for mine when placed my stone
POW1(a, b)	powerBoard1[PXY(a,b)]
	How many claim[] changes for mine when placed his stone
POW2(a, b)      powerBoard2[PXY(a,b)]
	(POW+ POW1) / 2

ISSAFE(a, b)    sList[gMap[stringIDs[PXY(a,b)]]].isSafe
ISDEAD(a, b)    sList[gMap[stringIDs[PXY(a,b)]]].isDead
	String on (a, b) is safe or dead.	

TRY_PLAY(a, b, c)       tryPlay(PCXY(a,b), c)
	Put a stone on (a, b) when 'c' is '1' put my stone, and
	when 'c' is '-1' put his stone.
	NOTE: When using TRY_PLAY you must call restoreState()

KILLABLE(a, b)          killable(PCXY(a,b), &xx, &yy)
SAVEABLE(a, b)          saveable(PCXY(a,b), &xx, &yy)
	String on (a,b) in pattern is killable or saveable.

PPOS(a,b)		Alternative Position

PERCENT(a)		(random() % 100) < a

Simple Pattern
==============

#Idnum Default_value
...........
..pattern..
...........

<Example>
#3 6
OX..
.@.O

- In this condition value of move on '@' is 6


Pattern with Code
=================
#104 8
|PT_SIDE
oX.*
O.@t
...t
> if (LIBC(0,1) < 3)
      value = 0;

- This pattern is for side.
  Defualt value is 8 but if Liberties of string on (0,1) is
less than 3 this move is useless(value = 0).


Complex Code
============
#Idnum default_value X_postion Y_postion

<Example>
#112 4 0 0
|PT_SIDE
*.Xo
t.Oo
> if (!ISSAFE(2,0) || LIBB(0,0) == 1)
    value = 0;
  else
    value += LIBB(0,0);

- This Pattern has no '@'.
  If you want check whose territory or forcefile of next move
indicate position like this.
  This Pattern's next move is (0, 0).

Alternative Position
====================
<Example>
#304 5.3
|PT_CORNER
......
..X...
....@.
......
......
> if (CLAIM(4, 2) > 4)
    value = 0;
  else if (PERCENT(30))
    PPOS(4,3);

- This patterns default position is '@'(4,2).
  Probability of alternative position (4,3) is 30%.

Tuning a new pattern
====================
After adding a new pattern you must test it works correctly.



ChangeLog
10/18/02:
	* (ADB) Badukiplus played its first game against Wallyplus, using the
    twogtp1or2.py Python script. 

10/18/02:
	* (ADB) Badukiplus played its first game against Gnu Go, using the
	  twogtp1or2.py Python script. First GTP V2 implementation finished.

10/13/02:
	* (ADB) First C++ version (compilation with gcc-3.2).

Below: previous ChangeLog and NEWS files

---------------------------------------------------------------------------

ChangeLog
=========

1999

1999. 1. 5	Baduki 0.1 released

1999. 1. 15	Baduki 0.2 released

2. 2	pattern.c. convert one pattern to eight direction.
2. 6	group renamed to string. include gID, gList
	intBoard wGIDMap : group id
	gen_wGIDMap(): get group id from scoreboard
2.11	pattern handling for side and corner
2.15	intBoard vsIDMap : ID for eye(territory)
2.22	Some endgame patterns added
	new variable: intboard powerBoard1
	revive function Getmove(lookForSave2)
	Strengthen endgame functions
	Baduki LOGO

1999. 3. 7	Baduki 0.2.1 released

3. 9	new program mkpat for pattern converting.

1999. 3. 10	Baduki 0.2.2 released

3.11	intBoard valueBoard for each moves value * 10

	intBoard myStones, hisStones for number of my/his stones of
	each point.

	thk_jump1() check for 2 point jump in 3 line.

3.12	in Blocking 1 Point extension and knight'move, two string must
	not in atari

	Capture opp. stones liberty is 2 when escaping.

	Avoid double atary

	Give bonus value save string kills undead string

	Think higher postion first in killing string

	In case a string can live with capture a opp. string or extend itself,
	extend if the liberty of string is 2 after capture

1999. 3. 13	Baduki 0.2.3 released

3. 16	Change gtk_signal_connect_object() to gtk_signal_connect()

	New function: get_moves_for_kill()
	More reliable for ladder.

	markDead() save killing/saving move in kx,ky/sx,sy
	and use this in thinking.

3. 17	killable() cosidering diagonal move first if necessary.

	Check a move is safe when the move is biggest.

	add counter in tryPlay(). save it sc/kc in string info

	New function: Getmove(fill_dame)

3. 18	Getmove(thk_jump1) check opp.'s weak stone.

3. 19	restoreState() is no necessary in pattern code.
	make alternative postion in pattern.  PPOS(x,y)
	
	Remake: Getmove(takeCorner) randomly select corner and postion

	markDead() mark undead if two adjacent strings are dead

1999. 3. 20	Baduki 0.2.4 released

3. 24	intBoard connectMap2 : opp's connection map

	New function: get_moves_for_save()
	
	reduce end games pattern's value

3. 28	Getmove(lookForKill) Kill dead stones if it can connect to safe stone

3. 30	get_moves_for_save/kill() Considering 1 point jump for kill/save.

	intBoard extMap[4] extension length from stone

	getMove() prevent move endangering safe stone

1999. 3. 31	Baduki 0.2.6 released

---------------------------------------------------------------------------

NEWS
====

0.2.6 (1999. 3. 31)
reduce end games pattern's value
Kill dead stones if it can connect to safe stone
Considering 1 point jump for kill/save.
prevent move endangering safe stone

0.2.5 (1999. 3. 23)
Fixed somg bugs.

0.2.4 (1999. 3. 20)
More reliable in Killing stones
Fill Dames.
Random move in take corner.

0.2.3 (1999. 3. 13)
In Blocking 1 Point extension and knight'move, two string must not in atari
  Capture opp. stones liberty is 2 when escaping.
Avoid double atary
Give bonus value save string kills undead string
Think higher postion first in killing string - good result in ladder.
In case a string can live with capture a opp. string or extend itself,
extend if the liberty of string is 2 after capture

0.2.2 (1999. 3. 10)
  You can easily add a new pattern.
  Read PATTERNS.

0.2.1 (1999. 3. 7)
  Group Information. (See MODE: wGIDMap and vsIDMap)
  Group in previous version renamed to String.
  Pattern converting and Matching.

0.2 (1999. 1. 16)
  Adopt automake/autoconf/gettext by Ryu, Changwoo <cwryu@adam.kaist.ac.kr>
	More portable and support NLS.
  Bugs in Undo/Redo fixed.
  A little more reliable in checking a stone is killable/saveable,
but spent much more time.

0.1 (1999. 1. 5)
  First public release.

In 4th FOST Cup (Oct 1998, Japan) Baduki ranked 21th.



badukiplus/patternsdat.h
 /* This file is automatially generated by mkpat from ./patterns.dat */
 /* Do not edit this file manually */

pattern_type pattern[] = {
  {10000, PT_CENTER, 3, 2, -1, -1, 0.000000,
	"O.O"
	"O.O"
  },
  {10001, PT_CENTER, 3, 2, -1, -1, 0.000000,
	"X.X"
	"X.X"
  },
  {1, PT_CENTER, 3, 3, 2, 1, 6.000000,
	".XO"
	"X.."
	"..."
  },
  {2, PT_CENTER, 4, 3, 2, 1, 6.000000,
	".OX*"
	"O..o"
	"...*"
  },
  {3, PT_CENTER, 4, 2, 1, 0, 6.000000,
	"...O"
	"OX.."
  },
  {4, PT_CENTER, 4, 2, 2, 0, 6.000000,
	"...O"
	"O.X*"
  },
  {5, PT_CENTER, 3, 2, 1, 1, 6.000000,
	"O.O"
	"x.X"
  },
  {6, PT_CENTER, 4, 2, 2, 1, 6.000000,
	"OO.o"
	"XX.."
  },
  {7, PT_CENTER, 3, 3, 1, 1, 0.000000,
	"*t*"
	"t.T"
	"*T*"
  },
  {8, PT_CENTER, 3, 2, 2, 1, 6.000000,
	".O."
	".X."
  },
  {9, PT_CENTER, 3, 3, 1, 1, 6.000000,
	"ooX"
	"..."
	"X.X"
  },
  {10, PT_CENTER, 3, 3, 1, 1, 2.000000,
	"TO*"
	"T.X"
	".tt"
  },
  {11, PT_CENTER, 4, 2, 1, 0, 3.100000,
	"...X"
	"XO.*"
  },
  {12, PT_CENTER, 4, 2, 2, 0, 3.100000,
	"...X"
	"X.O."
  },
  {13, PT_CENTER, 5, 3, 2, 1, 2.500000,
	"....."
	"X...X"
	"..O.."
  },
  {14, PT_CENTER, 5, 2, 2, 1, 2.000000,
	"....X"
	"X...."
  },
  {15, PT_CENTER, 2, 2, 1, 1, 2.000000,
	"XO"
	"O."
  },
  {16, PT_CENTER, 5, 3, 2, 1, 3.800000,
	"..X.O"
	"O...."
	"o...*"
  },
  {17, PT_CENTER, 2, 3, 1, 1, 6.000000,
	"Xx"
	"Ot"
	"Xx"
  },
  {18, PT_CENTER, 5, 3, 2, 2, 5.800000,
	"*.O.*"
	"X...x"
	"x...X"
  },
  {19, PT_CENTER, 3, 3, 1, 1, 3.000000,
	"oXo"
	"..."
	"oXo"
  },
  {20, PT_CENTER, 5, 4, 2, 1, 6.000000,
	"....."
	"O...O"
	"....."
	"**X**"
  },
  {21, PT_CENTER, 3, 3, 1, 1, 3.000000,
	"O.*"
	"..X"
	"*.O"
  },
  {22, PT_CENTER, 5, 3, 2, 2, 6.000000,
	"..O.X"
	"X...."
	"....."
  },
  {23, PT_CENTER, 5, 3, 2, 1, 6.000000,
	"..X.O"
	"O...."
	"....."
  },
  {24, PT_CENTER, 2, 3, 0, 1, 5.500000,
	".T"
	"TO"
	".X"
  },
  {25, PT_CENTER, 4, 3, 1, 2, 6.000000,
	"*OO*"
	"X..X"
	"o..o"
  },
  {26, PT_CENTER, 3, 3, 1, 1, 3.000000,
	"O.X"
	"..."
	"*.O"
  },
  {27, PT_CENTER, 3, 2, 1, 1, 5.000000,
	"*XO"
	"O.."
  },
  {104, PT_SIDE, 4, 3, 2, 1, 5.900000,
	"o..."
	"O..t"
	"oX.*"
  },
  {105, PT_SIDE, 5, 3, 3, 0, 5.400000,
	"o...t"
	"O...s"
	"*Xx**"
  },
  {106, PT_SIDE, 4, 3, 2, 0, 5.000000,
	"o..t"
	"O..s"
	"*Xx*"
  },
  {107, PT_SIDE, 3, 3, 1, 0, 4.000000,
	"o.t"
	"O.s"
	"*Xx"
  },
  {108, PT_SIDE, 4, 3, 3, 1, 6.000000,
	"...."
	".O.."
	"OXtt"
  },
  {109, PT_SIDE, 2, 3, 1, 1, 6.000000,
	"Tt"
	"Tt"
	"OX"
  },
  {110, PT_SIDE, 3, 3, 2, 1, 8.000000,
	"..."
	".X."
	".O."
  },
  {111, PT_SIDE, 4, 2, 2, 0, 2.000000,
	"S..."
	"SOXs"
  },
  {112, PT_SIDE, 4, 2, 0, 0, 4.000000,
	"t.Oo"
	"*.Xo"
  },
  {113, PT_SIDE, 4, 3, 0, 2, 6.000000,
	"...."
	".OXX"
	"..O."
  },
  {114, PT_SIDE, 3, 3, 0, 1, 7.000000,
	"*.x"
	".XO"
	"oOX"
  },
  {115, PT_SIDE, 3, 3, 1, 2, 6.000000,
	"..*"
	".OX"
	"..O"
  },
  {116, PT_SIDE, 5, 3, 2, 1, 6.000000,
	"o.o.*"
	"O...O"
	"OXXO*"
  },
  {117, PT_SIDE, 4, 4, 2, 1, 5.000000,
	"...."
	"...."
	"O..x"
	"**Xx"
  },
  {118, PT_SIDE, 5, 4, 3, 1, 6.000000,
	"....."
	"....."
	"O...*"
	"o..Xx"
  },
  {119, PT_SIDE, 4, 4, 1, 1, 4.000000,
	"...."
	"...."
	"X..o"
	"x.Oo"
  },
  {120, PT_SIDE, 5, 4, 3, 2, 6.000000,
	"....."
	"....."
	"O..t."
	"....."
  },
  {121, PT_SIDE, 4, 3, 2, 1, 4.000000,
	"...."
	"O..s"
	"*.X*"
  },
  {122, PT_SIDE, 3, 3, 2, 2, 3.000000,
	"..*"
	"..X"
	"**."
  },
  {123, PT_SIDE, 3, 3, 1, 1, 2.000000,
	"T.t"
	"T.t"
	"O.X"
  },
  {124, PT_SIDE, 4, 3, 2, 1, 4.800000,
	"T..t"
	"T..t"
	"O..X"
  },
  {125, PT_SIDE, 4, 2, 2, 0, 5.600000,
	"O..o"
	"..XO"
  },
  {126, PT_SIDE, 4, 2, 3, 0, 3.000000,
	"oO.."
	"*X.t"
  },
  {127, PT_SIDE, 4, 2, 1, 0, 5.600000,
	"O..o"
	"*X.O"
  },
  {128, PT_SIDE, 3, 2, 1, 0, 6.000000,
	"o.o"
	"OXO"
  },
  {129, PT_SIDE, 3, 3, 2, 1, 6.000000,
	"O.."
	"XO."
	"*X."
  },
  {130, PT_SIDE, 3, 2, 2, 1, 6.000000,
	".O."
	"*Xt"
  },
  {131, PT_SIDE, 2, 3, 0, 1, 3.000000,
	".."
	".O"
	".X"
  },
  {132, PT_SIDE, 3, 3, 1, 1, 5.000000,
	"..."
	"o.o"
	"OXO"
  },
  {133, PT_SIDE, 4, 4, 2, 3, 6.000000,
	"o..."
	"o..."
	"O..."
	"...."
  },
  {134, PT_SIDE, 3, 4, 0, 3, 3.000000,
	"***"
	"SSo"
	"TTO"
	"tt."
  },
  {135, PT_SIDE, 3, 4, 0, 2, 2.500000,
	"SSo"
	"TTO"
	"..."
	"ttX"
  },
  {136, PT_SIDE, 4, 4, 1, 2, 6.000000,
	"...."
	"...."
	"O..O"
	".X.."
  },
  {301, PT_CORNER, 5, 5, 3, 2, 6.000000,
	"....."
	"....."
	"..X.."
	"...O."
	"....."
  },
  {302, PT_CORNER, 5, 5, 2, 2, 6.000000,
	"....."
	"...X."
	"....."
	"...O."
	"....."
  },
  {303, PT_CORNER, 6, 6, 2, 5, 6.500000,
	"......"
	"......"
	".....X"
	"...O.."
	"......"
	"......"
  },
  {304, PT_CORNER, 6, 5, 4, 2, 5.300000,
	"......"
	"......"
	"......"
	"..X..."
	"......"
  },
  {305, PT_CORNER, 6, 6, 3, 3, 5.300000,
	"......"
	"......"
	"..X..."
	"......"
	"......"
	"......"
  },
  {305, PT_CORNER, 6, 6, 3, 3, 6.000000,
	"......"
	"......"
	"...X.."
	"....O."
	"......"
	"......"
  },
  {306, PT_CORNER, 6, 6, 3, 2, 5.300000,
	"......"
	"......"
	"......"
	"......"
	"..X..*"
	"....**"
  },
  {307, PT_CORNER, 6, 6, 3, 2, 5.300000,
	"......"
	"......"
	"......"
	".....*"
	"...X.*"
	"...x**"
  },
  {308, PT_CORNER, 2, 2, 1, 0, 0.500000,
	".."
	"XO"
  },
  {309, PT_CORNER, 6, 6, 2, 4, 5.300000,
	"......"
	"......"
	"...O.."
	"......"
	"...X.."
	"......"
  },
  {310, PT_CORNER, 6, 6, 2, 3, 5.300000,
	"......"
	"......"
	"..O..."
	"...X.."
	"......"
	"......"
  },
  {311, PT_CORNER, 6, 6, 5, 2, 5.250000,
	"......"
	"......"
	"......"
	"...X.."
	"......"
	"......"
  },
  {312, PT_CORNER, 4, 4, 2, 2, 5.100000,
	"...."
	"...O"
	"...."
	"...X"
  }
};

